solutions/ojuz-knapsack.mdx

```diff
@@ -0,0 +1,174 @@
+
+## Implemention
+
+**Time Complexity:** $\mathcal{O}(S^2 \cdot \log S)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+using std::pair;
+
+int main() {
+	int limit;
+	int type_num;
+	std::cin >> limit >> type_num;
+	
+	std::map<int, vector<pair<int, int>>> by_weight;
+	for (int t = 0; t < type_num; t++) {
+		int value;
+		int weight;
+		int amt;
+		std::cin >> value >> weight >> amt;
+		if (weight <= limit && 0 < amt) {
+			by_weight[weight].push_back({value, amt});
+		}
+	}
+	
+	/*
+	 * best[i][j] contains the most value we can
+	 * get using j weight and the first i weight types
+	 */
+	vector<vector<long long>> best(
+		by_weight.size() + 1, vector<long long>(limit + 1, INT32_MIN)
+	);
+	best[0][0] = 0;
+	int at = 1;
+	for (auto& [w, items] : by_weight) {
+		// sort items from most value to least value
+		std::sort(items.begin(), items.end(), std::greater<pair<int, int>>());
+		for (int i = 0; i <= limit; i++) {
+			best[at][i] = best[at - 1][i];
+			int copies = 0;
+			int type_at = 0;
+			int curr_used = 0;
+			long long profit = 0;
+			// go through as many items until we run out of items or usable weight
+			while ((copies + 1) * w <= i && type_at < items.size()) {
+				copies++;
+				profit += items[type_at].first;
+				if (best[at - 1][i - copies * w] != INT32_MIN) {
+					best[at][i] = std::max(
+						best[at][i],
+						best[at - 1][i - copies * w] + profit
+					);
+				}
+				
+				curr_used++;
+				if (curr_used == items[type_at].second) {
+					curr_used = 0;
+					type_at++;
+				}
+			}
+		}
+		at++;
+	}
+	cout << *std::max_element(
+		best.back().begin(),
+		best.back().end()
+	) << endl;
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+<Warning>
+Due to Java's (relatively) slow speed, this solution will often TLE on the last subtask.
+</Warning>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class knapsack {
+	public static void main(String[] args) throws IOException {
+		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
+		StringTokenizer initial = new StringTokenizer(read.readLine());
+		int limit = Integer.parseInt(initial.nextToken());
+		int typeNum = Integer.parseInt(initial.nextToken());
+		HashMap<Integer, ArrayList<int[]>> byWeight = new HashMap<>();
+		for (int t = 0; t < typeNum; t++) {
+			StringTokenizer item = new StringTokenizer(read.readLine());
+			int value = Integer.parseInt(item.nextToken());
+			int weight = Integer.parseInt(item.nextToken());
+			int amt = Integer.parseInt(item.nextToken());
+			if (weight <= limit && 0 <= amt) {
+				if (!byWeight.containsKey(weight)) {
+					byWeight.put(weight, new ArrayList<>());
+				}
+				byWeight.get(weight).add(new int[] {value, amt});
```

Comment:
Can you make an `Item` class that has instance variables `value` and `amt` instead? That way it's easier to follow along.

Reply:
~~don't wanna make a constructor and all that~~
alright i'll do it

Reply:
haven't done it yet
but i think the constant factor might make it tle?

Reply:
![image](https://user-images.githubusercontent.com/55369003/128538704-ca6e84ee-0690-4fee-adcf-99dd65e86b0a.png)
yup sadge